### PR TITLE
Implements CLI tool for rule management

### PR DESCRIPTION
**Pull request by falun**
## Description
This PR introduces `rulesctl` as a command line tool to interface with the filecustomrule method of encoding/configuring the query rules. It initially implements list, remove-rule, and a simplistic version of add-rule.

This makes a few minor refactors but no behavioral to the rules/tablet server code. 

### Usage Examples
```
[~/go/src/vitess.io/vitess] (master)
falun@vitess-build-host $ ./rulesctl -f ~/rules.json list
[
  {
    "Description": "This is a test rule",
    "Name": "basic_rule",
    "Query": ".*some_bad_query_identifier.*",
    "Plans": [
      "Select",
      "Insert"
    ],
    "TableNames": [
      "table1",
      "table2"
    ],
    "Action": "FAIL"
  },
  {
    "Description": "A second basic rule",
    "Name": "basic_rule 2",
    "Query": ".*",
    "Plans": [
      "Select"
    ],
    "Action": "FAIL"
  }
]

falun@vitess-build-host $ ./rulesctl -f ~/rules.json list --names-only
[
  "basic_rule",
  "basic_rule 2"
]

falun@vitess-build-host $ ./rulesctl -f ~/rules.json remove-rule -n "basic_rule 2"

falun@vitess-build-host $ ./rulesctl -f ~/rules.json list --names-only
[
  "basic_rule"
]

falun@vitess-build-host $ ./rulesctl -f ~/new-rules.json add-rule -n 'example_rule' -a fail -e 'a new rule' -p select

falun@vitess-build-host $ ./rulesctl -f ~/new-rules.json list
[
  {
    "Description": "a new rule",
    "Name": "example_rule",
    "Query": "",
    "Plans": [
      "Select"
    ],
    "Action": "FAIL"
  }
]
```

## Related Issue(s)
No open issues, relevant channel in community slack #feat-query-block.

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

## Deployment Notes
N/a

## Impacted Areas in Vitess
No impact to Vitess deployments. This is a command line tool to facilitate safely constructing config files.

### CODE REVIEW

File: go/cmd/rulesctl/main.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+func mustPrintJSON(obj interface{}) {
+	enc, err := json.MarshalIndent(obj, "", "  ")
+	if err != nil {
+		log.Fatalf("Unable to marshal object: %v", err)
+	}
+	fmt.Printf("%v\n", string(enc))
+}
+
+func mustWriteJSON(obj interface{}, path string) {
```
Comment (ajm188) [inline go/cmd/rulesctl/main.go:152]: tioli (it's in an awkward namespacing for reuse, etc) there's a `MarshalJSON` in [vtctldclient/cli](https://github.com/vitessio/vitess/blob/master/go/cmd/vtctldclient/cli/json.go#L37) that does the same indenting

File: go/vt/vttablet/customrule/filecustomrule/filecustomrule.go [hunk 2]
```diff
@@ -53,23 +53,35 @@ func NewFileCustomRule() (fcr *FileCustomRule) {
 	return fcr
 }
 
+// ParseRules will construct a Rules object based on a file path. On any error
+// nil and that error will be returned. A log will be printed to capture the
```
Comment (ajm188) [inline go/vt/vttablet/customrule/filecustomrule/filecustomrule.go:57]: This was confusing to me at first, in that I thought maybe you dropped some words. What do you think of:

> In the case of any error, it returns nil and that error.

Or maybe just adding a comma so "on any error nil" => "on any error, nil and that error"

**Review by ajm188 — COMMENTED** [review]
This is amazing!! I have some minor nitpicks around wording, and then a couple code suggestions, which I think are totally fine to take-or-leave!

**Comment by derekperkins** [comment]
I honestly have no idea what a filecustomrule is, so I don’t know how useful this is, but this goes against what was roughly agreed upon two meetups ago as a goal to reduce the number of binaries #7471. 

If we felt like this functionality was important enough, maybe we could stub out the framework for a new vitess binary and let this be the first subcommand? Maybe we could merge this experimentally before doing that, with the understanding that it would likely be the first binary removed as we take an incremental approach to binary consolidation?

File: go/cmd/rulesctl/main.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+		Use:  "rulesctl",
+		Args: cobra.NoArgs,
+		PreRun: func(cmd *cobra.Command, args []string) {
+			tmp := os.Args
+			os.Args = os.Args[0:1]
+			flag.Parse()
+			os.Args = tmp
+		},
+	}
+	rootCmd.Run = func(_ *cobra.Command, _ []string) { rootCmd.Help() }
```
[the remark this answers is not in the document]
Comment (setassociative) [reply go/cmd/rulesctl/main.go:30]: ohhhh, clever

File: go/cmd/rulesctl/main.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+	removeCmd.MarkFlagRequired("name")
+
+	removeCmd.Run = func(cmd *cobra.Command, args []string) {
+		rules := getRules()
+		if deleted := rules.Delete(removeOptName); deleted == nil {
+			fmt.Printf("No rule found: '%v'", removeOptName)
+			return
+		}
+
+		if removeOptDryRun {
```
[the remark this answers is not in the document]
Comment (falun) [reply go/cmd/rulesctl/main.go:78]: :brain: 
:white_check_mark:

[the remark this answers is not in the document]
Comment (falun) [reply go/cmd/rulesctl/main.go:78]: Ugh, I love this but dealing with a file/writer means I can't then use ioutil.WriteFile and it wouldn't actually simplify the calling code bc it'd still need to branch on the dry run code.

Going to leave this as is.

File: go/cmd/rulesctl/main.go (hunk 1 above)
Comment (falun) [reply go/cmd/rulesctl/main.go:152]: going to leave it for now; it's simple enough i don't mind the reproduction enough to have the slightly weird package dep

**Comment by falun** [comment]
Made requested changes from @ajm188.

@derekperkins While I'm interested in the outcome of the single-binary conversation I don't want to litigate/block as part of this PR. So what I took the approach of making everything well factored to be pulled into a universal binary but also build independently as-is. 

I think we can also use this approach as a model of what it would look like for the sub-commands off a global `vitess`; the parent binary's main would basically be:

```go
package main

import (
	"log"

	rulesctl "vitess.io/vitess/go/cmd/rulesctl/cmd"
)

func main() {
	rootCmd := &cobra.Command{Name: "vitess"
	if err := rootCmd.Execute(); err != nil {
		log.Printf("%v", err)
	}
}
```


func Main() *cobra.Command {
	rootCmd := &cobra.Command{
		Use:  "rulesctl",
		Args: cobra.NoArgs,
		PreRun: func(cmd *cobra.Command, args []string) {

**Comment by derekperkins** [comment]
I think that looks awesome as a rough framework, and keeping the binary itself well isolated is a good example for how we should isolate the rest of the binaries, regardless of the uni-binary outcome.

**Comment by deepthi** [comment]
> I think that looks awesome as a rough framework, and keeping the binary itself well isolated is a good example for how we should isolate the rest of the binaries, regardless of the uni-binary outcome.

+1
@ajm188 can you re-review and approve/merge if all feedback has been addressed?

**Review by ajm188 — APPROVED** [review]
lg2m!

**Pull request merged**